### PR TITLE
Allow toggling of feature_flag_called

### PR DESCRIPTION
**Pull request by timgl**
## Changes
...

## Checklist
- [ ] Tests for new code (if applicable)
- [ ] TypeScript definitions (module.d.ts) updated and in sync with library exports (if applicable)

### CODE REVIEW

**Comment by timgl** [comment]
Closes #100

File: src/__tests__/featureflags.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { PostHogFeatureFlags } from '../posthog-featureflags'
+import { _ } from '../utils'
+
+fdescribe('featureflags', () => {
+    given('instance', () => ({
+        get_property: (key) => {
+            if (key === '$override_feature_flags') return false
```
Comment (macobo) [inline src/__tests__/featureflags.js:6]: Proposal:
```
        get_property: (key) => given.properties['key']

  given('properties', () => ({ $override_feature_flags: false, $active_feature_flags: ['beta-feature'] })
```

File: src/posthog-featureflags.js [hunk 2]
```diff
@@ -51,14 +51,17 @@ export class PostHogFeatureFlags {
      *     if(posthog.isFeatureEnabled('beta-feature')) { // do something }
      *
      * @param {Object|String} prop Key of the feature flag.
+     * @param {Bool} prop Whether to send a .
```
Comment (macobo) [inline src/posthog-featureflags.js:54]: You a word.

```
 @param {Bool} send_event (Optional) Whether to send $feature_flag_called event to posthog
```

File: src/posthog-featureflags.js [hunk 3]
```diff
@@ -51,14 +51,17 @@ export class PostHogFeatureFlags {
...
      */
-    isFeatureEnabled(key) {
+    isFeatureEnabled(key, send_event = true) {
```
Comment (macobo) [inline src/posthog-featureflags.js:56]: Would it make sense to make this into ```isFeatureEnabled(key, options = {})```?

This way the calling code will make a bit more sense: `.isFeatureEnabled('feature', false)` is rather opaque, `.isFeatureEnabled('feature', { send_event: false })` gives an idea of what's going on

**Review by macobo — CHANGES_REQUESTED** [review]
Don't forget to update typing!

:+1: for adding tests.

Longer-term, do we want this to be a top-level configuration option?

File: src/__tests__/featureflags.js (hunk 1 above)
Comment (timgl) [reply src/__tests__/featureflags.js:6]: Don

File: src/posthog-featureflags.js (hunk 2 above)
Comment (timgl) [reply src/posthog-featureflags.js:54]: Done

File: src/posthog-featureflags.js (hunk 3 above)
Comment (timgl) [reply src/posthog-featureflags.js:56]: Yep good point, done

File: src/__tests__/featureflags.js (hunk 1 above)
Comment (macobo) [reply src/__tests__/featureflags.js:6]: :+1: don't forget to push

File: src/module.d.ts [hunk 4]
```diff
@@ -692,8 +697,9 @@ declare namespace posthog {
          *     if(posthog.isFeatureEnabled('beta-feature')) { // do something }
          *
          * @param {Object|String} prop Key of the feature flag.
+         * @param {Object|String} options (optional) If {send_event: false}, we won't send an $feature_flag_call event to PostHog.
          */
-        static isFeatureEnabled(key: string): boolean
+        static isFeatureEnabled(key: string, options: posthog.isFeatureEnabledOptions): boolean
```
Comment (macobo) [inline src/module.d.ts:702]: This needs to be optional: `options?: posthog.isFeatureEnabledOptions`

Comment (macobo) [reply src/module.d.ts:702]: We can inline isFeatureEnabledOptions - `options?: { send_event?: boolean }`

File: src/posthog-featureflags.js [hunk 5]
```diff
@@ -51,14 +51,17 @@ export class PostHogFeatureFlags {
...
      */
-    isFeatureEnabled(key) {
+    isFeatureEnabled(key, options = {}) {
         if (!this.getFlags()) {
             console.warn('isFeatureEnabled for key "' + key + '" failed. Feature flags didn\'t load in time.')
             return false
         }
         let response = this.getFlags().indexOf(key) > -1
-        this.instance.capture('$feature_flag_called', { $feature_flag: key, $feature_flag_response: response })
+        if (options.send_event || Object.keys(options).indexOf('send_event') === -1) {
```
Comment (macobo) [inline src/posthog-featureflags.js:62]: `!('send_event' in options)`

**Review by macobo — CHANGES_REQUESTED** [review]

**Comment by razor-x** [comment]
Helpful change :)

I wrote some UNTESTED code to check how my implementation would need to change to use this

First, update the function that gets the feature flags. I can't really control how often this function is called since it lives in a react hook and could be called in many components. I can only ensure it is called at least 1 time per second.

```ts
  const getFlags = () =>
    featureFlags.map(
      (k) =>
        !!posthog.isFeatureEnabled(k, {
          send_event: willSendEvent(k)
        })
    )
```

Now implement `willSendEvent` to throttle the events to 5/min.


```ts
const minEventInterval = 5 * 60 * 1000
const sentEventTimes = {}
const willSendEvent = (k) => {
  const now = Date.now()
  const lastEventTime = sentEventTimes[k]

  const isLongerThanMinInterval =
    lastEventTime && lastEventTime + minEventInterval < now

  if (!lastEventTime || isLongerThanMinInterval) {
    sentEventTimes[k] = now
    return true
  }

  return false
}
```

**Review by macobo — APPROVED** [review]
:rocket:

**Comment by mariusandra** [comment]
I'm obviously late with this comment, but wouldn't it make sense to default this to `false` instead of `true`? Then the users who really need the extra events to track the feature flag usage would get them, everyone else would be saved from this "noise"?

Just my 0.02 SGD, feel free to ignore.

**Comment by timgl** [comment]
You can do some really powerful things with this data (ie do cohorts of "how do people retain if they've seen feature flag A rather than B") so I'd rather default to capturing this data. we should monitor though and see if people end up using this at all or not

**Comment by razor-x** [comment]
Yes I was going to recommend the same. In this case, the feature has already been released, so it may not be correct to disable it for existing users.

However, for a future change like this that introduces a new event to a method that did not have it before, the default should be disabled, at least on the first minor version that has it. That way users can opt in and assess impact. The feature can be enabled by default on a follow up major release.

**Comment by razor-x** [comment]
Will there be a new release soon :)

**Comment by macobo** [comment]
@razor-x sorry for the delay, 1.6.0 is on NPM now :+1:

**Pull request merged**